File: dotnet/src/Pactman.NonprofitCheckPlus/PactmanClient.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using Pactman.NonprofitCheckPlus.Configuration;
using Pactman.NonprofitCheckPlus.Exceptions;
using Pactman.NonprofitCheckPlus.Http;

using Pactman.NonprofitCheckPlus.Internal;

namespace Pactman.NonprofitCheckPlus
{
    /// <summary>
    /// The Pactman Nonprofit Check Plus client.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Server-side use only. The API key is a private credential; do not construct this
    /// client anywhere its configuration is shipped to an end user.
    /// </para>
    /// <code>
    /// using var client = new PactmanClient(Environment.GetEnvironmentVariable("PACTMAN_API_KEY"));
    /// var result = await client.Nonprofits.CheckAsync("41-1787097");
    /// </code>
    /// <para>
    /// Build one client per process and share it — it is safe to use concurrently. Each
    /// instance carries its own connection pool and throttle state, so constructing one
    /// per request throws both away and, on a self-owned client, leaks sockets through
    /// <c>TIME_WAIT</c>. In an ASP.NET Core app, register it as a singleton and hand it an
    /// <c>IHttpClientFactory</c> client through <see cref="PactmanClientOptions.HttpClient"/>.
    /// </para>
    /// </remarks>
    public sealed class PactmanClient : IDisposable
    {
        private readonly ClientConfig _config;
        private readonly HttpClient _httpClient;
        private readonly bool _disposeHttpClient;

        private bool _disposed;

        /// <summary>Constructs a client for the production API with default options.</summary>
        /// <param name="apiKey">
        /// Your Pactman API key. Load it from the environment or a secret manager; never
        /// commit it, and never ship it to an end user.
        /// </param>
        /// <exception cref="PactmanConfigurationException">The key is missing or blank.</exception>
        public PactmanClient(string? apiKey)
            : this(new PactmanClientOptions { ApiKey = apiKey })
        {
        }

        /// <summary>Constructs a client from a full set of options.</summary>
        /// <param name="options">The options to resolve.</param>
        /// <exception cref="ArgumentNullException"><paramref name="options"/> is <see langword="null"/>.</exception>
        /// <exception cref="PactmanConfigurationException">
        /// A missing or blank API key, an unknown environment, a malformed base URL, or a
        /// nonsensical numeric option.
        /// </exception>
        public PactmanClient(PactmanClientOptions options)
            : this(options, null)
        {
        }

        internal PactmanClient(PactmanClientOptions options, TransportHooks? hooks)
        {
            if (options is null)
            {
                throw new ArgumentNullException(nameof(options));
            }

            var key = ConfigResolver.ApiKey(options.ApiKey);
            _config = ConfigResolver.Resolve(options);

            if (options.HttpClient is null)
            {
                // Deadlines are enforced per attempt with a cancellation token, so the
                // client's own timeout is stood down rather than racing ours.
                _httpClient = new HttpClient { Timeout = System.Threading.Timeout.InfiniteTimeSpan };
                _disposeHttpClient = true;
            }
            else
            {
                _httpClient = options.HttpClient;
                _disposeHttpClient = options.DisposeHttpClient;
            }

            // The key lives only inside this delegate. It is not a field of the client,
            // the configuration or the transport, so it appears in no property, no
            // ToString(), no ToDictionary() and nothing that serializes those.
            string Credential() => "Bearer " + key;

            Nonprofits = new NonprofitsResource(new Transport(Credential, _config, _httpClient, hooks));
        }

        /// <summary>Nonprofit lookups.</summary>
        public NonprofitsResource Nonprofits { get; }

        /// <summary>The resolved base URL every request is sent to.</summary>
        public string BaseUrl => _config.BaseUrl;

        /// <summary>
        /// The named environment in use, or <see langword="null"/> when an explicit base URL
        /// was given.
        /// </summary>
        public PactmanEnvironment? Environment => _config.Environment;

        /// <summary>The resolved timeout per attempt.</summary>
        public TimeSpan Timeout => _config.Timeout;

        /// <summary>The resolved retry policy.</summary>
        public RetryOptions Retry => _config.Retry;

        /// <summary>
        /// A redacted view of the configuration.
        /// </summary>
        /// <remarks>
        /// The API key is not a property of this object and never appears here, in
        /// <see cref="ToString"/>, or in anything that serializes either.
        /// </remarks>
        /// <returns>The configuration's fields, with the key redacted.</returns>
        public IReadOnlyDictionary<string, object?> ToDictionary()
        {
            var fields = new FieldMap(_config.ToDictionary())
            {
                ["apiKey"] = "[redacted]",
            };

            return fields;
        }

        /// <inheritdoc />
        public override string ToString() =>
            string.Format(CultureInfo.InvariantCulture, "PactmanClient({0})", _config.BaseUrl);

        /// <summary>
        /// Releases the HTTP client this instance owns.
        /// </summary>
        /// <remarks>
        /// A client supplied through <see cref="PactmanClientOptions.HttpClient"/> is left
        /// alone unless <see cref="PactmanClientOptions.DisposeHttpClient"/> asked otherwise,
        /// so disposing this does not tear down a pool you share.
        /// </remarks>
        public void Dispose()
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;

            if (_disposeHttpClient)
            {
                _httpClient.Dispose();
            }
        }
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/PactmanClient.java
package org.pactman.nonprofitcheckplus;

import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import org.pactman.nonprofitcheckplus.config.ClientConfig;
import org.pactman.nonprofitcheckplus.config.PactmanClientOptions;
import org.pactman.nonprofitcheckplus.config.ConfigResolver;
import org.pactman.nonprofitcheckplus.config.RetryOptions;
import org.pactman.nonprofitcheckplus.http.Transport;
import org.pactman.nonprofitcheckplus.http.TransportHooks;

/**
 * Entry point for the SDK.
 *
 * <p>Server-side use only. The API key is a private credential; do not construct
 * this client in a mobile binary, a desktop application, or any other context
 * where the artifact is shipped to a user.
 *
 * <p>Thread-safe and designed to be long-lived: build one per application and
 * share it, so connections and the rate limiter are shared too.
 *
 * <pre>{@code
 * try (PactmanClient client = new PactmanClient(System.getenv("PACTMAN_API_KEY"))) {
 *     SingleCheckResult result = client.nonprofits().check("41-1787097");
 * }
 * }</pre>
 *
 * <p>{@link #close()} only shuts down the thread pool the async methods use, and
 * only if they were used; a client that is never closed leaks nothing when no
 * async call was made.
 */
public final class PactmanClient implements AutoCloseable {

    private final ClientConfig config;
    private final Transport transport;
    private final NonprofitsResource nonprofits;

    /** Created on first async use, so the common case allocates no threads. */
    private volatile ExecutorService executor;

    private final Object executorLock = new Object();

    /**
     * Builds a client with every default, from an API key.
     *
     * @param apiKey your Pactman API key.
     */
    public PactmanClient(String apiKey) {
        this(PactmanClientOptions.of(apiKey));
    }

    /**
     * Builds a client.
     *
     * @param options the configuration. The API key is required.
     * @throws org.pactman.nonprofitcheckplus.exceptions.PactmanConfigurationException
     *         if the options are unusable.
     */
    public PactmanClient(PactmanClientOptions options) {
        this(options, null);
    }

    /**
     * Builds a client with a substituted clock and random source.
     *
     * <p>The seam the test suite uses to assert on retry delays without waiting
     * them. It is not covered by semantic versioning; do not depend on it.
     *
     * @param options the configuration. The API key is required.
     * @param hooks   the clock and random source, or {@code null} for the real ones.
     */
    public PactmanClient(PactmanClientOptions options, TransportHooks hooks) {
        ConfigResolver.Resolved resolved = ConfigResolver.resolve(options);
        this.config = resolved.config();
        this.transport = new Transport(resolved.apiKey(), this.config, hooks);
        this.nonprofits = new NonprofitsResource(this.transport, this::runAsync);
    }

    /**
     * Nonprofit lookups.
     *
     * @return the resource.
     */
    public NonprofitsResource nonprofits() {
        return nonprofits;
    }

    /**
     * The resolved base URL every request is sent to.
     *
     * @return the base URL.
     */
    public String baseUrl() {
        return config.baseUrl();
    }

    /**
     * The named environment in use.
     *
     * @return the environment, or {@code null} when an explicit base URL was given.
     */
    public PactmanEnvironment environment() {
        return config.environment();
    }

    /**
     * The resolved timeout.
     *
     * @return the timeout in milliseconds.
     */
    public long timeoutMs() {
        return config.timeoutMs();
    }

    /**
     * The resolved retry policy.
     *
     * <p>Start from this when overriding one setting for a single request:
     * {@code client.retry().toBuilder().maxRetries(0).build()}.
     *
     * @return the policy.
     */
    public RetryOptions retry() {
        return config.retry();
    }

    /**
     * A redacted view of the configuration.
     *
     * <p>The API key is not a property of this object and never appears here or
     * in {@link #toString()}.
     *
     * @return the settings, in a stable order.
     */
    public Map<String, Object> toMap() {
        return config.toMap();
    }

    /**
     * Releases the thread pool the async methods use.
     *
     * <p>In-flight async calls are interrupted. Nothing else is held: the HTTP
     * client, if this SDK built one, is released with the client itself.
     */
    @Override
    public void close() {
        ExecutorService running = executor;

        if (running != null) {
            running.shutdownNow();
        }
    }

    @Override
    public String toString() {
        return "PactmanClient(" + config.baseUrl() + ")";
    }

    private void runAsync(Runnable work) {
        executor().execute(work);
    }

    private ExecutorService executor() {
        ExecutorService running = executor;

        if (running != null) {
            return running;
        }

        synchronized (executorLock) {
            if (executor == null) {
                executor = newExecutor();
            }

            return executor;
        }
    }

    /**
     * An unbounded pool of daemon threads that retire when idle.
     *
     * <p>The work is blocking I/O, so it must not run on the common
     * ForkJoinPool — a batch of lookups there would starve every other parallel
     * stream in the application. Daemon threads keep a client the caller forgot
     * to close from holding the JVM open.
     */
    private static ExecutorService newExecutor() {
        AtomicLong counter = new AtomicLong();

        return new ThreadPoolExecutor(
                0,
                Integer.MAX_VALUE,
                60L,
                TimeUnit.SECONDS,
                new SynchronousQueue<Runnable>(),
                runnable -> {
                    Thread thread = Executors.defaultThreadFactory().newThread(runnable);
                    thread.setName("pactman-nonprofit-check-" + counter.incrementAndGet());
                    thread.setDaemon(true);

                    return thread;
                });
    }
}
